web/frontend/libs/legal_document_search.js

```diff
@@ -0,0 +1,164 @@
+    { val: "vt", name: "Vermont" },
+    { val: "wash", name: "Washington" },
+    { val: "wis", name: "Wisconsin" },
+    { val: "w-va", name: "West Virginia" },
+    { val: "wyo", name: "Wyoming" },
+  ];
+
+/**
+ * Accepts a query for a legal document and returns the result list
+ *
+ * @param {string} query the query to pass long to the search
+ * @param {array} allSources all available sources
+ * @param {?number} sourceId the id of the only source to query
+ * @param {?string} jurisdiction code for the state/jurisdiction to limit the search
+ * @param {?string} beforeDate YYYY-MM-DD date string
+ * @param {?string} afterDate YYYY-MM-DD date string *
+ *
+ * @returns {array} the list of results by source
+ */
+export const search = async (
+  query,
+  allSources,
+  sourceId,
+  jurisdiction,
+  beforeDate,
+  afterDate
+) => {
+  const api = url.url("search_using");
+
+  const sources = [];
+  const sourceDetail = allSources.filter((s) =>
+    sourceId ? s.id === sourceId : true
+  );
+
+  let order = 0; // Sources will come back ordered in "priority order", which we want to retain
+
+  for (const { id, name } of sourceDetail) {
+    const url =
+      api({ sourceId: id }) +
+      "?" +
+      new URLSearchParams({
+        q: query,
+        jurisdiction: jurisdiction || "",
+        before_date: beforeDate || "",
+        after_date: afterDate || "",
+      });
+    sources.push({ url, id, name, order });
+    order += 1;
+  }
+  return await Promise.all(
+    sources.map(async (source) => {
+      const { url, id, name, order } = source;
+      return fetch(url)
+        .then((r) => r.json())
+        .then((r) => {
+          const { results } = r;
+          return results.map((row) => {
+            row.id = row.id.toString(); // normalize IDs from the API to strings
+            return {
+              name,
+              sourceId: id,
+              sourceOrder: order,
+              ...row,
+            };
+          });
+        });
+    })
+  );
+};
+/**
+ * Given a casebook and optional section ID, add a legal document
+ * from the identified source to the casebook and return the location of the 
+ * resource in the book.
+ * 
+ * @param {string} casebookId the ID of the casebook
+ * @param {?string} sectionId the optional ID of the section in which to nest it
+ * @param {string} sourceRef the foreign identifier of the legal document
+ * @param {number} sourceId the local identifier of the source (e.g. CAP)
+ * 
+ * @returns {Object}
+ */
+export const add = async (casebookId, sectionId, sourceRef, sourceId) => {
+  const api = url.url("legal_document_resource_view");
+
+  const resp = await fetch(api({ casebookId }), {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+      "X-CSRF-Token": get_csrf_token(),
+    },
+    body: JSON.stringify({
+      source_id: sourceId,
+      source_ref: sourceRef,
+      section_id: sectionId,
+    }),
+  });
+  const body = await resp.json();
```

Comment:
I am not sure how the API you're interacting with behaves so this might be irrelevant, but given how frequently I forget about this myself, it's probably worth mentioning: `fetch()` doesn't throw when it encounters HTTP != 2XX. 

With the code above, if your API returns something like: 
```javascript
HTTP 500 
{ "error": "Something terrible happened." }
```

`fetch()` would still resolve, so would `response.json()`, and I think this function would return something like.
```javascript
{
  resourceId: undefined, 
  redirectUrl: undefined, 
  sourceRef: "value of sourceRef"
}
```

I tend to add a status code check that throws if status != 200 on whatever comes out of `fetch` for that reason.

Cheers.

Reply:
How am I supposed to know that when it's only in **very noticeable bold type** in MDN?? 

<img width="727" alt="image" src="https://user-images.githubusercontent.com/19571/231156927-a3d19baa-e02c-44fe-ae4e-f4fe69579adf.png">

Thanks, I did miss this. It's our API so we'll know from Sentry that it threw on the server side, but agreed the client should recover better especially since it's possible the client just passed some unexpected parameters (which does cause a 500, ask me how I know) and the rest of the site might be working fine. I'll address this in a fast follow since I want to get this on staging for Catherine to review.